### PR TITLE
discoverd2: Implement leader election

### PR DESCRIPTION
**Pull request by titanous**
This is an incremental patch for #145.

### CODE REVIEW

File: discoverd2/server/state_test.go [hunk 1]
```diff
@@ -193,40 +219,237 @@ func (StateSuite) TestSetService(c *C) {
...
+	c.Assert(state.GetLeader("a"), DeepEquals, third)
+
+	// set with same instance and another instance triggers no events
+	fourth := fakeInstance()
+	fourth.Index = 5
+	state.SetService("a", []*Instance{fourth, third})
+	assertNoEvent(c, events)
+	c.Assert(state.GetLeader("a"), DeepEquals, third)
+
+	// set with same instance and lower selects a new leader
```
Comment (heavenlyhash) [inline discoverd2/server/state_test.go]: s/lower/lower index/

File: discoverd2/server/state.go [hunk 2]
```diff
@@ -199,93 +196,185 @@ type State struct {
...
+		instances: make(map[string]*Instance),
+	}
+}
+
+type service struct {
+	// instance ID -> instance
+	instances map[string]*Instance
+
+	leaderID     string
+	leaderIndex  uint64
```
Comment (heavenlyhash) [inline discoverd2/server/state.go]: Perhaps a quick comment that this is usually constrained to being `>0` as per the `Instance.Index` field, with a sentinel value for zero -> unset?

**Comment by titanous** [comment]
@heavenlyhash Comments addressed.

**Comment by heavenlyhash** [comment]
LGTM

**Pull request merged**